piptools/scripts/compile.py

```diff
@@ -359,10 +363,7 @@ def cli(
             if not only_build_deps:
                 constraints.extend(metadata.requirements)
                 if all_extras:
-                    if extras:
-                        msg = "--extra has no effect when used with --all-extras"
-                        raise click.BadParameter(msg)
-                    extras = metadata.extras
+                    extras += metadata.extras
```

Comment:
Why does this have `+=`? How is it related to just moving the check to another location as the PR title advertises?

Reply:
Actually, it doesn't make a difference, does it? It's verified that `extras` and `all_extras` aren't combined beforehand, so we could as well remove it.

```suggestion
                    extras = metadata.extras
```

Reply:
It is required. Otherwise, only the extras from the last iteration (the last `src_file`) will be used.

I agree this detail is missing from the description and title. I will add it now.